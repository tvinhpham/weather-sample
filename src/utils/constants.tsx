//export const API_URL = 'https://www.metaweather.com';

//export const API_URL = 'https://cors-anywhere.herokuapp.com/https://www.metaweather.com';
export const API_URL = 'https://meta-weather.vercel.app';
