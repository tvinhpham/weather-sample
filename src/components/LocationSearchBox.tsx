import React from 'react';
import axios from 'axios';
import { FunctionComponent, HTMLAttributes, useState } from 'react';
import { AsyncTypeahead } from 'react-bootstrap-typeahead';
import { useQuery } from 'react-query';

import 'react-bootstrap-typeahead/css/Typeahead.min.css';
import InputGroup from 'react-bootstrap/InputGroup';
import Alert from 'react-bootstrap/Alert';
import { API_URL } from '../utils/constants';

const searchLocations = async ({ queryKey }: any) => {
  let [, query] = queryKey;

  if (!query) {
    return [];
  }

  const { data } = await axios.get(`${API_URL}/api/location/search/?query=${query}`);
  return data;
};

const LocationSearchBox: FunctionComponent<
  {
    onLocationSelected?: (loc: any) => void;
  } & HTMLAttributes<HTMLElement>
> = ({ onLocationSelected, ...props }) => {
  const [searchText, setSearchText] = useState('');

  const { isLoading, isError, data } = useQuery(['locations', searchText], searchLocations);

  const onSearch = (query: string) => {
    setSearchText(query);
  };

  return (
    <div {...props}>
      <InputGroup {...props}>
        <InputGroup.Text>
          <img
            src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8
            /9hAAAAAXNSR0IArs4c6QAAAQ1JREFUOE+d078rhlEYxvHPSwabMluVgfJHGAw2WcQkpchk4FWKjUFJJJQSSdnY/QdKySL
            /gcIiKXTred4eT8+Pek+d5Vx33/u6r3NOQ/nqwHeF/ic1cgW9WEIfvtCFHxzhtgiWBfRjCyt4yBR3o4l3bOYhKSA6XWEar
            yW2V3GHm6yeAiYTy5cVM0eTM0wUAWLGBXzUhLaHZbyldamDU0zVJY51nOA5D9jFGl5qIMeYzzpNHYxgADsVgJ5Ej6BbK3uN
            59jAYwEkHlV038Z9GSA6HOIaF/hMCgeT8YYSwH4ZIM6j0zhG0ZnsJ0T6YzjAIiKzwqdcdxEzeUj+L9QBQv8HaQcQkDkMY7Z
            dQMvpL1evLxHmgIo1AAAAAElFTkSuQmCC"
            alt="Search icon"
          />
        </InputGroup.Text>

        <AsyncTypeahead
          placeholder="Search"
          data-testid="searchLocationBox"
          // @ts-ignore
          labelKey="title"
          filterBy={() => true}
          id="locationBox"
          options={data || []}
          onChange={([loc]) => onLocationSelected?.(loc)}
          minLength={2}
          renderMenuItemChildren={(option: any) => (
            <div key={option.woeid}>
              <span>{option.title}</span>
            </div>
          )}
          onSearch={onSearch}
          isLoading={isLoading}
        />
      </InputGroup>
      {isError && (
        <Alert key="locationSearchError" variant="danger">
          Unexpected error occurs: Cannot connect to server
        </Alert>
      )}
    </div>
  );
};

export default LocationSearchBox;
