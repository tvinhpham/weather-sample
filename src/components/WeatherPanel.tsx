import React from 'react';
import axios from 'axios';
import { FunctionComponent, HTMLAttributes } from 'react';
import { useQuery } from 'react-query';

import 'react-bootstrap-typeahead/css/Typeahead.min.css';
import DayForecast from './DayForecast';
import Spinner from 'react-bootstrap/Spinner';
import Alert from 'react-bootstrap/Alert';
import { API_URL } from '../utils/constants';

const getLocationDetails = async ({ queryKey }: any) => {
  let [, locId] = queryKey;

  if (!locId) {
    return null;
  }

  const { data } = await axios.get(`${API_URL}/api/location/${locId}/`);
  return data;
};

const WeatherPanel: FunctionComponent<
  {
    locationId?: number;
  } & HTMLAttributes<HTMLDivElement>
> = ({ locationId, ...props }) => {
  const { isLoading, isError, data } = useQuery(['locationDetails', locationId], getLocationDetails);

  if (isLoading) {
    return <Spinner animation="grow" data-testid="loadingForcast" />;
  }

  if (isError) {
    return (
      <Alert key="forecastPnError" variant="danger">
        Unexpected error occurs: Cannot connect to server
      </Alert>
    );
  }

  return (
    <div {...props}>
      {
        // @ts-ignore
        data?.consolidated_weather?.slice(0, 5)?.map((fc: any, index: number) => (
          <DayForecast
            key={locationId + '_' + index}
            data={fc}
            className="col-12 col-sm-6 col-md-4 col-lg-2 p-2 p-md-1 p-lg-0"
          />
        ))
      }
    </div>
  );
};

export default WeatherPanel;
