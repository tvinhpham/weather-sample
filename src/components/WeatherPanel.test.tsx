import React from 'react';
import renderer from 'react-test-renderer';
import LocationSearchBox from './LocationSearchBox';
import { queryClient } from '../utils/helper';
import { QueryClientProvider } from 'react-query';
import WeatherPanel from './WeatherPanel';

it('renders with loading expectation', () => {
  // @ts-ignore
  const searchBox = renderer
    .create(<QueryClientProvider client={queryClient}><WeatherPanel locationId={2487956} /></QueryClientProvider>)
    .toJSON();
  expect(searchBox).toMatchSnapshot();
});

