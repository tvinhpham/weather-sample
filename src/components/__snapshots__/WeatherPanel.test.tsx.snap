// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with loading expectation 1`] = `
<div
  className="spinner-grow"
  data-testid="loadingForcast"
/>
`;
